lib/main.ts

```diff
@@ -0,0 +1,144 @@
+ * automatically prints errors that occurred to the console.
+ *
+ * Use the {@link exit} operation form within to halt program execution
+ * immediately and initiate shutdown.
+ *
+ * The behavior of `main` is slightly different depending on the environment it
+ * is running in.
+ *
+ * ### Deno, Node
+ *
+ * When running within Deno or Node, any error which reaches `main` causes the
+ * entire process to exit with an exit code of `1`.
+ *
+ * Additionally, handlers for `SIGINT` are attached to the
+ * process, so that sending an exit signal to it causes the main task
+ * to become halted. This means that hitting CTRL-C on an Effection program
+ * using `main` will cause an orderly shutdown and run all cleanup code.
+ *
+ * > Warning! do not call `Deno.exit()` on Deno or `process.exit()` on Node
+ * > directly, as this will not gracefully shutdown. Instead, use the
+ * > {@link exit} operation.
+ *
+ * ### Browser
+ *
+ * When running in a browser, The `main` operation gets shut down on the
+ * `unload` event.
+ */
+
+export async function main(
+  op: (args: string[]) => Operation<void>,
+): Promise<void> {
+  let hardexit = (_status: number) => {};
+
+  let result = await run(() => action<Exit>(function* (resolve) {
+    // action will return shutdown immediately upon resolve, so stash
+    // this function in the exit context so it can be called anywhere.
+    yield* ExitContext.set(resolve);
+
+    try {
+      let interrupt = () => resolve({ status: 130, signal: "SIGINT " });
+      return yield* withHost({
+        *deno() {
+          hardexit = (status) => Deno.exit(status);
+          try {
+            Deno.addSignalListener("SIGINT", interrupt);
+            yield* op(Deno.args.slice());
+          } finally {
+            Deno.removeSignalListener("SIGINT", interrupt);
+          }
+        },
+        *node() {
+          //@ts-expect-error type-checked by Deno, run on Node
+          hardexit = (status) => global.process.exit(status);
+          try {
+            //@ts-expect-error type-checked by Deno, run on Node
+            process.on("SIGINT", interrupt);
+            //@ts-expect-error type-checked by Deno, run on Node
+            yield* op(global.process.args.slice());
+          } finally {
+            //@ts-expect-error this runs on Node
+            process.off("SIGINT", interrupt);
+          }
+        },
+        *browser() {
+          try {
+            self.addEventListener("unload", interrupt);
+            yield* op([]);
+          } finally {
+            self.removeEventListener("unload", interrupt);
+          }
+        },
+      });
+    } catch (error) {
+      resolve({ status: 1, error });
+    }
+
+    yield* exit(0);
+  }));
+
+  if (result.message) {
+    if (result.status === 0) {
+      console.log(result.message);
+    } else {
+      console.error(result.message);
+    }
+  }
+
+  if (result.error) {
+    console.error(result.error);
+  }
+
+  hardexit(result.status);
```

Comment:
I'm curious, if `effection` properly shuts the task tree down, why do we need a `hardexit`?

Reply:
@neurosnap It's in the case where the exit status is non-zero you have to actually have to transmit the exit code to any caller of the process. I guess we don't need it when the exit code is `0`.

I was also toying with the idea of making main an operation so that it could be fully embedded, but it means you would have to have some more boilerplate:

```ts
let result = await run(main(function*(args) {
  ///  your program here
}));
if (result.status !== 0) {
  Deno.exit(result.status);
}

the advantage of the above is that your main method is fully embeddable, and you could have mains embedded with other mains within mains...